### DISCUSSION TITLE
Can I reload a TreeControl's list of items?

### DISCUSSION THREAD

**Discussion by MasterKale** [Textualize/textual#151] [Q&A] [answered]
I've been tinkering with Textual to build with a TUI that can be used to view details about hardware devices connected to my computer. On the left is the list of devices in `TreeControl` and the right is info about an item in the tree. So far I've managed to get things working to the point that selecting an item in the tree populates the details pane to the right of it:

https://user-images.githubusercontent.com/5166470/137944222-8724c37a-5c52-4a0b-91f0-599fb0c24124.mov

What's stumped me now is, how can I display a new list of items in the tree? I want to offer a "Reload" command that will rescan for hardware devices and then display the new list in place of the old. Unfortunately all I've been able to manage is adding additional items, with the unfortunate side effect of the app crashing with an `Exception: Wrong channel` if you happen to click an item that existed in the tree prior to adding more:

https://user-images.githubusercontent.com/5166470/137944589-0451ccea-0644-408e-b486-a56d8bbb004c.mov

I'm attempting to reload the items like this, with a method on my subclass of `TreeControl` that the app can pass new devices to:

```py
async def on_reload_authenticators(
    self, new_authenticators: List[AuthenticatorViewModel]
) -> None:
    self.authenticators = new_authenticators
    for auth_view_model in self.authenticators:
        tree_data = AuthenticatorListData(authenticator=auth_view_model)
        await self.add(self.root.id, auth_view_model.aaguid, tree_data.dict())
    self.refresh(layout=True)
```

But this clearly isn't the complete formula.

What I _think_ I want is the ability to specify async `render()` methods so that I can declare `self.authenticators` as a `Reactive()` and hopefully the tree list Just Updates when I specify a new value for `self.authenticators`. However the issue with that is that `render()` seems to want to be synchronous, and I can't reconcile that with the fact that I need to await calls to `TreeControl.add()`.

Has anyone tackled this before? Will said he was inspired by VueJS for some of Textual's inner workings, and I'm coming at this from a predominantly web dev background; I was hoping Textual might emulate some of the magic of the "one-way data flow" of front end SPA frameworks with automatic re-renderings. In this case `TreeControl` seems to need a bit more manual manipulation...

**Comment by chelnak** [comment, marked as answer]
Hey @MasterKale , this is how I do it:

https://github.com/chelnak/jenkins-tui/blob/016958abe639e97daa02d63535fcdfde7f1b97cd/src/jenkins_tui/app.py#L100

I've got a refresh method in the works that will remove the need to init a new version of the tree.

**Reply by MasterKale** [reply]
This worked perfectly, thank you! This has the added benefit of always triggering `on_mount()` logic, too, which keeps things clean. 

Maybe I was approaching this too much like web front end 😅
